### ISSUE TITLE
Ability to change which tab an Activity appears in (or no tab at all)

### ISSUE THREAD

**Issue by avaidyam** [BIDMCDigitalPsychiatry/LAMP-platform#165]
The Activity create/update UI should have a drop-down selector to choose from: 
1. `default` (`null`): The Activity icon will go to the default tab (i.e. surveys in assess and tips in learn).
2. `learn`: The Activity icon will go to the Learn tab.
3. `assess`: The Activity icon will go to the Assess tab.
4. `manage`: The Activity icon will go to the Manage tab.
5. `prevent`: The Activity icon will go to the Prevent tab (above the Prevent graphs).
6. `none`: This Activity icon will be entirely hidden from the UI **but notifications will still fire and data should still be collected.**

**Comment by sarithapillai8** [comment]
@avaidyam 
We discussed about this earlier and this tab information was expected to be saved as part of activity object in a key "tab". The key tab does not exist now and to support this we need to change LAMP-server activity modal and some other related places. Please confirm on this. Or do you prefer any other method?

**Comment by avaidyam** [comment]
@sarithapillai8 Please add the field `category` (see below type definition) to both `Activity` and `ActivitySpec`. 

```typescript
type Tab = 'learn' | 'assess' | 'manage' | 'prevent'
type Activity {
    category: Tab[] | null
}
type ActivitySpec {
    category: Tab[] | null
}
```

- If `Activity.category` is missing (`undefined`), it must be treated as "use default category" (which is the same behavior if `=== null`). 
- Valid options for the `category` array are `learn`, `assess`, `manage`, `prevent`. 
- If the array is empty, it should appear in no tabs in the UI. 
- If the array has a single item, such as `learn`, it must show up in the Learn tab in the UI. 
- It is **NOT** permitted to have duplicates, such as `['learn', 'learn']` (we do not need to add any checks for this in the API or UI code; see below).
- **For the current implementation, if `activity.category.length > 0`, ONLY use `activity.category[0]`.** (Do not allow showing an activity in multiple tabs for now.)
- For the current implementation, filter out all invalid tab names using `filter()` or `find()` (i.e. `let tab: string | undefined = activity?.category?.find(x => ['learn', 'assess', 'manage', 'prevent'].contains(x))?.first`) on the client/UI-side before choosing a tab.
- If `Activity.category` does not have a valid category (`undefined` or `null`), repeat this lookup for `ActivitySpec`. 

```
if activity.category.first exists:
    set tab = ...
else if get_activity_spec(activity.spec).category.first exists:
    set tab = ...
else: # for backwards compatibility
    if activity.spec == 'lamp.tips':
        set tab = 'learn'
    if activity.spec in manage_activities_list:
        set tab = 'manage'
    else: 
        set tab = 'assess'
```

This will require **minor changes only** change in `LAMP-server` and `LAMP-js`, and **no database-side changes at all.** 

// cc @lukeoftheshire

**Comment by Linoy339** [comment]
@avaidyam . Please find 
[https://github.com/BIDMCDigitalPsychiatry/LAMP-js/pull/3](https://github.com/BIDMCDigitalPsychiatry/LAMP-js/pull/3)

We have merged Lamp-server changes which has been merged to staging for testing.
Please find the changes we have done in staging:
[https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/162](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/162)

**Comment by sarithapillai8** [comment]
@avaidyam 
Could you please update this PR and make it available so that the feature can be tested in staging.
https://github.com/BIDMCDigitalPsychiatry/LAMP-js/pull/3

**Comment by sarithapillai8** [comment]
@avaidyam 
Could you please confirm that -  if the tab 'none' is selected, do we need to hide it from feed or can we show that as this is available in notification?

**Comment by avaidyam** [comment]
> Could you please confirm that - if the tab 'none' is selected, do we need to hide it from feed or can we show that as this is available in notification?

It should still be visible on the feed if it is scheduled, but it should not tappable unless within the notification window.

**Comment by sarithapillai8** [comment]
@avaidyam 
Could you please release the LAMP-js changes? The changes in LAMP-dashboard are not yet updated on staging because of this.

**Comment by avaidyam** [comment]
I've released it.

**Comment by sarithapillai8** [comment]
@avaidyam 
I am getting this error in LAMP-dashboard build which I fixed in local by changing node_modules lamp-core directly and building  : 

/usr/src/app/src/components/Researcher/ActivityList/ActivityMethods.ts
TypeScript error in /usr/src/app/src/components/Researcher/ActivityList/ActivityMethods.ts(967,23):
Property 'category' does not exist on type 'ActivitySpec'.  TS2339

    965 | export async function getDefaultTab(spec) {
    966 |   let activitySpec = await getActivitySpec(spec)
  > 967 |   if (!!activitySpec?.category) {
        |                       ^
    968 |     return activitySpec?.category[0]
    969 |   } else {
    970 |     if (

**Comment by avaidyam** [comment]
Did you bump the version to `2021.11.15`? Otherwise if you are experiencing any issues you may need to communicate with @Linoy339. I was under the impression that all prerequisite changes were made before I was asked to publish a production release, so I'm not sure how to help.

**Comment by sarithapillai8** [comment]
@avaidyam 
Ok So I have to update the version number. Thanks.

**Comment by sarithapillai8** [comment]
@SuruPa00 This is updated in production.

**Comment by SuruPa00** [comment]
@sarithapillai8 this is working almost as intended, but right now when multiple tabs are selected for which the activity should go under, it only goes under the first one

**Comment by sarithapillai8** [comment]
@SuruPa00 
Please see the comment from @avaidyam  https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/165#issuecomment-965362762:
For the current implementation, if activity.category.length > 0, ONLY use activity.category[0]. (Do not allow showing an activity in multiple tabs for now.)

**Comment by avaidyam** [comment]
Yes, now that the initial functionality is working it makes sense to add multiple tabs in.

**Comment by sarithapillai8** [comment]
@SuruPa00 @avaidyam  
The changes with multiple tabs updated in production.

**Comment by sarithapillai8** [comment]
@avaidyam @SuruPa00 
We have a confusion with None option. Suppose there is an activity with tabs selected - None and Assess. How is the working expected there? Can we just disable other tab selections on selecting 'None' tab in the activity creation?

**Comment by SuruPa00** [comment]
@sarithapillai8 if none is selected, it should not be in any tab so yes, go ahead and make it so if none is selected it cannot appear in other tabs even if multiple are selected

**Comment by avaidyam** [comment]
> Valid options for the category array are learn, assess, manage, prevent.

@sarithapillai8 Are you sure you implemented this feature correctly? It seems to me like you didn't read the implementation instructions I provided at all. (See above.)

**Comment by sarithapillai8** [comment]
@avaidyam 
The values saved are learn, assess, manage, prevent. But there is an additional option 'None' as per requirement. It is given as a checkbox and an empty string saving for this selection, which is used for checking none. Do you mean to just make an empty array when checking 'None'?

**Comment by avaidyam** [comment]
Yes, that is what is stated: 

> If the array is empty, it should appear in no tabs in the UI.

**Comment by sarithapillai8** [comment]
The default tab will be selected by default and this 5 options available for checking. So I think I misunderstood this. The None checkbox can be removed and when the user manually uncheck any default selection/previous selections, it should be considered as none. I can update this change by tomorrow, if this is correct.
![image](https://user-images.githubusercontent.com/63710347/145231672-a4aa95be-b666-4f1e-8f42-e40bc41b5f30.png)

**Comment by avaidyam** [comment]
In a multiple state control like a group of checkboxes, you shouldn't add an explicit "none" option as that is logically incorrect. If no checkboxes are selected, it already implies "none" - which is why an empty array means such.  

Please add a separate checkbox "Customize which Tab this Activity appears in". If this checkbox is disabled, the `category` property should be set to `null` and the category selector picker (in the screenshot) should be disabled/un-tappable. If it is enabled, the category property is an empty array by default, and multiple options can be selected from the picker.

**Comment by sarithapillai8** [comment]
Ok got it. I can make the changes ASAP.

**Comment by sarithapillai8** [comment]
@avaidyam 
Changes are updated in production.
